service/sharddistributor/handler/executor.go

```diff
@@ -57,16 +57,74 @@ func (h *executor) Heartbeat(ctx context.Context, request *types.ExecutorHeartbe
 
 	heartbeatTime := h.timeSource.Now().UTC()
 	mode := h.migrationConfiguration.GetMigrationMode(request.Namespace)
+
+	// Handle draining executor separately
+	if request.Status == types.ExecutorStatusDRAINING {
+		return h.handleDrainingHeartbeat(ctx, request, heartbeatTime, mode)
+	}
+
+	return h.handleActiveHeartbeat(ctx, request, previousHeartbeat, assignedShards, heartbeatTime, mode)
+}
+
+// handleDrainingHeartbeat handles heartbeats from executors that are shutting down.
+// It clears their shard assignments, so they can be reassigned to other executors.
+func (h *executor) handleDrainingHeartbeat(
+	ctx context.Context,
+	request *types.ExecutorHeartbeatRequest,
+	heartbeatTime time.Time,
+	mode types.MigrationMode,
+) (*types.ExecutorHeartbeatResponse, error) {
+	h.logger.Info("Executor is draining, clearing shard assignments",
+		tag.ShardNamespace(request.Namespace),
+		tag.ShardExecutor(request.ExecutorID))
+
+	if err := h.storage.DeleteExecutors(ctx, request.Namespace, []string{request.ExecutorID}, store.NopGuard()); err != nil {
+		h.logger.Error("Failed to delete draining executor's assignments",
+			tag.ShardNamespace(request.Namespace),
+			tag.ShardExecutor(request.ExecutorID),
+			tag.Error(err))
+		// Continue anyway - the leader will eventually reassign the shards
+	}
```

Comment:
I think we need a "set executors draining" if we delete the assignments we will have shards that are not assigned to anything. This causes problems with processes that assume all shards are assigned. We should probably remove the delete executors completely. 

I think (please verify) that the leader loop is watching the status of the executors, so writing to ETCD that the executer is draining should trigger the leader to rebalance.